[REVIEW] Converting all Estimator Constructors to Keyword Arguments

mdemoret-nv opened the pull request:
Closes #3373 

This mimics sklearn's positional argument deprecation message (and is named differently to prevent mixups) but is automatically added to any class that inherits from `cuml.Base` (similar to other `CumlArray` decorators that are automatically added to `Base` class). When the descriptor is automatically added, it assumes all arguments should be keyword-only (to catch accidentally forgetting to specify keyword only arguments). This can easily be overridden by manually adding the decorator if some arguments need to be positional and some keyword. For example, the `OneVsRestClassifier` estimator takes a positional estimator to wrap so the `__init__` function looks like:
```python
@_deprecate_pos_args(version="0.20")
def __init__(self,
             estimator,
             *args,
             handle=None,
             verbose=False,
             output_type=None):
```

Notes: 

1. This only updates `__init__` functions but could be applied to any of the public functions in the API. 
2. No tests were added because validation is performed on module load and will generate an assert error if developers have used it incorrectly

JohnZed commented on python/cuml/internals/base_helpers.py:
```diff
@@ -125,14 +125,69 @@ def _wrap_attribute(class_name: str,
 
+def _check_and_wrap_init(attribute, **kwargs):
+
+    from cuml.common.input_utils import _deprecate_pos_args
```
Could be imported at module level

mdemoret-nv replied:
This cannot be imported at module level. Circular import. `common` -> `internals` -> `common`. I'm open to a better location or structure to avoid this if you have any ideas.

dantegd replied:
This form of local import is surprisingly expensive, but also technically `_deprecate_pos_args` is not processing inputs, so it would be just as good placed in `internals` and then this can be avoided altogether

mdemoret-nv replied:
Moved this. Resolved

JohnZed commented on python/cuml/internals/base_helpers.py:
```diff
@@ -103,8 +104,7 @@ def _wrap_attribute(class_name: str,
     # Skip items marked with autowrap_ignore
-    if ("__cuml_is_wrapped" in attribute.__dict__
-            and attribute.__dict__["__cuml_is_wrapped"]):
+    if (attribute.__dict__.get("__cuml_is_wrapped", False)):
```
Nitpick: _deprecate_pos_args.FLAG_NAME is always used through a constant and this flag, which seems conceptually almost the same, is used through a literal

mdemoret-nv replied:
I couldnt do exactly the same thing (There is no common decorator class for `__cuml_is_wrapped` to be attached to) but I did create the constant `CUML_WRAPPED_FLAG`. Let me know if this resolves this

JohnZed replied:
Yeah, looks good to me now.

JohnZed reviewed (commented):
Looks very cool! I had only minor questions and comments, can wrap up quickly. If this is ready to go, can you remove the WIP label?

mdemoret-nv commented on python/cuml/cluster/kmeans.pyx:
```diff
@@ -263,11 +263,13 @@ class KMeans(Base,
                  init='scalable-k-means++', n_init=1, oversampling_factor=2.0,
                  max_samples_per_batch=1<<15, output_type=None):
-        super(KMeans, self).__init__(handle, verbose, output_type)
+        super().__init__(handle=handle,
```
This actually reminded me that I forgot to this for the dask classes. They should all be updated in this PR now.

dantegd reviewed (changes requested):
Just had a quick comment

mdemoret-nv commented on python/cuml/common/input_utils.py:
```diff
@@ -609,3 +613,70 @@ def sparse_scipy_to_cp(sp, dtype):
     return cupyx.scipy.sparse.coo_matrix((v, (r, c)), sp.shape)
+
+
+class _deprecate_pos_args:
```
Here are the numbers. Before this PR:
```python
>>> timeit.repeat('cuml.DBSCAN()', 'import cuml', number=100000)
[1.6759078549985134, 0.41760462100137374, 0.4198731740034418, 0.4204189170013706, 0.4161926310007402]
```
After this PR:
```python
>>> timeit.repeat('cuml.DBSCAN()', 'import cuml', number=100000)
[2.557721577999473, 0.7288761550007621, 0.7315405129993451, 0.7273198049988423, 0.7297004659994855]
```
I was surprised that the impact would be so big (75% longer) but at the same time this is only adding 3 usec per instantiation which is incredibly small. I'm not sure what we could change in the function to make it run faster.

mdemoret-nv commented on python/cuml/common/type_utils.py:
```diff
@@ -13,8 +13,27 @@
+
+# Use _F as a type variable for decorators. See:
+# https://github.com/python/mypy/pull/8336/files#diff-eb668b35b7c0c4f88822160f3ca4c111f444c88a38a3b9df9bb8427131538f9cR260
+_F = typing.TypeVar("_F", bound=typing.Callable[..., typing.Any])
```
Resolved.

mdemoret-nv commented:
@dantegd and @JohnZed I've incorporated your feedback into this PR and merged upstream. However, with the changes to the `BaseEstimator` class in preprocessing, the number of updates I had to make was pretty large. Probably warrants a quick re-review to make sure I didnt miss anything.

To address both of your comments about the local import, I've moved the `_deprecate_pos_args` class to `cuml.internal`. So this local import isnt needed anymore.

JohnZed commented on python/cuml/_thirdparty/sklearn/utils/validation.py:
```diff
@@ -41,42 +41,42 @@
-        kwargs.update({k: arg for k, arg in zip(sig.parameters, args)})
-        return f(**kwargs)
-    return inner_f
+# def _deprecate_positional_args(f):
```
Delete commented code?

mdemoret-nv replied:
Removed the commented code. Resolved.

JohnZed commented:
@gpucibot merge
